### ISSUE TITLE
np.linalg.svd hangs

### ISSUE THREAD

**Issue by cancan101** [numpy/numpy#10742]
Repost of https://github.com/nipy/nibabel/issues/609:
On certain machines, the following code reliable hangs (never returns):
```python
R = np.array([[1.        , 0.        ,         np.nan],
       [0.        , 0 ,         np.nan],
       [ 0.        ,  0,         np.nan]])
np.linalg.svd(R)
```

Numpy has been installed from wheel (numpy 1.14.2, Python 3.5.2):
```
np.show_config()
blas_mkl_info:
  NOT AVAILABLE
lapack_mkl_info:
  NOT AVAILABLE
blas_opt_info:
    language = c
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    define_macros = [('HAVE_CBLAS', None)]
openblas_info:
    language = c
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    define_macros = [('HAVE_CBLAS', None)]
lapack_opt_info:
    language = c
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    define_macros = [('HAVE_CBLAS', None)]
blis_info:
  NOT AVAILABLE
openblas_lapack_info:
    language = c
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    define_macros = [('HAVE_CBLAS', None)]

$ x86report
x86cpu report
-------------
brand            : Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
vendor           : GenuineIntel
model (display)  : 79
family (display) : 6
model            : 15
family           : 6
extended model   : 4
extended family  : 0
stepping         : 1
processor type   : 0
signature        : 263921
MMX              : True
3DNow!           : True
SSE              : True
SSE2             : True
SSE3             : True
SSSE3            : True
SSE4.1           : True
SSE4.2           : True
supports AVX     : True
supports AVX2    : True
```
This issue happens on all these openblas kernels:
```
dunnington
penryn
core2
nehalem
sandybridge
prescott
Haswell
```
/CC @matthew-brett

**Comment by charris** [comment]
Yep, that looks to be an openblas problem. Did you have the same problem with NumPy 1.14.1?

**Comment by charris** [comment]
Looks like a NumPy 1.14.3 is in our future :(

**Comment by cancan101** [comment]
I see this issue going back as far as numpy 1.9.0. I didn't test before that as there were no wheels for 3.5

**Comment by charris** [comment]
@cancan101 Anything special about the OS, libc?  What machines does it *not* hang on?

**Comment by cancan101** [comment]
Running inside a Docker image (`tensorflow/tensorflow:1.6.0-devel-gpu-py3`):

```
$ uname -a
Linux node015-jupyter-20180216-131819 4.4.0-1050-aws #59-Ubuntu SMP Tue Jan 30 19:57:10 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ cat /etc/issue
Ubuntu 16.04.3 LTS \n \l
```

**Comment by cancan101** [comment]
This demonstrates the issue: `docker run --rm tensorflow/tensorflow:1.6.0-devel-gpu-py3 python3 -c 'import numpy as np; R = np.array([[1.        , 0.        ,         np.nan],       [0.        , 0 ,         np.nan], [ 0.        ,  0,         np.nan]]) ;np.linalg.svd(R)'`

**Comment by matthew-brett** [comment]
On that container - same hang on Python 2.  Same hang building from source against apt-get installed openblas:

```
apt-get update
apt-get install -y libopenblas-dev
pip uninstall -y numpy
pip install --no-binary numpy numpy
```

Expected LinAlgError when compiling numpy without installed lapack, or with `libatlas-dev`, so it does appear to be specific to OpenBLAS.

Same error in bare Ubuntu 16.04 container:

```
docker pull ubuntu:16.04
docker run -ti --rm ubuntu:16.04 /bin/bash
```

then:

```
apt-get update
apt-get install -y python3 curl
curl -LO https://bootstrap.pypa.io/get-pip.py
python3 get-pip.py
pip install numpy
python3 -c 'import numpy as np; R = np.array([[1. , 0. , np.nan], [0. , 0 , np.nan], [ 0. , 0, np.nan]]) ;np.linalg.svd(R)'
```

**Comment by matthew-brett** [comment]
I think we'll need to build a C / OpenBLAS test case, test it against current OpenBLAS, and report it to the maintainers.   I'll put an issue up now in case it's obvious to them what to do.

**Comment by matthew-brett** [comment]
With help from the OpenBLAS team - we just need the latest develop branch of OpenBLAS.  I've built that, changes to use it at https://github.com/MacPython/numpy-wheels/blob/daily/config.sh#L3

**Comment by ahaldane** [comment]
I'm thinking of preparing the 1.14.3 numpy release now. Are we aiming to try and fix this soon?

**Comment by matthew-brett** [comment]
Sorry to be slow - do you mean that you've tested a wheel from our daily builds, and it doesn't cause this hang?

**Comment by charris** [comment]
No. I missed that we don't have a test for this. There are also reported errors with threading in OpenBLAS `0.2.20`, see #11046.

**Comment by matthew-brett** [comment]
The other problem is that we are getting scipy test failures on Windows 32 bit, with the versions of OpenBLAS that fix this problem, and we haven't got to the bottom of why.  Or even half way down.

https://github.com/scipy/scipy/issues/8759

**Comment by matthew-brett** [comment]
No - I'm afraid not - I'm a bit stuck as to what to do to debug.

**Comment by charris** [comment]
I'm going to release 1.14.4 soon. Would it be correct to use the same OpenBLAS build as the previous versions? Currently in `config.sh`

* daily: OPENBLAS_VERSION="0.2.20-343-g2c7392f0"
* v1.14.x: OPENBLAS_VERSION=0.2.18

The windows versions are the same in both daily and v1.14.x.

If there is no advantage to `0.2.20` I would just as soon leave v1.14.x as is.

**Comment by mattip** [comment]
@charris I am +1 for leaving 0.2.18 for the 1.14 releases even though it seems there are issues with threading. For 1.15, what is your timeline for release? OpenBlas 0.3.0 was [released](https://github.com/xianyi/OpenBLAS/releases) May 23.

**Comment by charris** [comment]
For 1.15 I am waiting on the PRs milestoned (a word?)  for it, but I may kick a few down the road just to get it out.

**Comment by mattip** [comment]
So I would be +1 for keeping 1.15 with OpenBlas 0.2.20, and waiting for feedback if the multiple threading issues get better or worse.

**Comment by mattip** [comment]
@matthew-brett what would you recommend for 1.15 which will be released soon?

**Comment by matthew-brett** [comment]
I'm just running Windows builds for OpenBLAS v0.3.0 - let's see how they go.  Otherwise, I guess we stick with what we know, and accept the infinite hangs that this OP reported.

**Comment by matthew-brett** [comment]
It looks like Sourceforge just went down : I'm getting curl download errors, and pages such as the blog are redirecting to a page with "The sourceforge.net website is temporarily in static offline mode."

**Comment by matthew-brett** [comment]
OK - all built now.  OpenBLAS v0.3.0 is passing all the Scipy tests, unlike earlier versions of OpenBLAS that I tried, that had this fix.   I don't know why, and I suspect we have not seen the last of these Scipy failures, but I think we can update to OpenBLAS v0.3.0:

* https://ci.appveyor.com/project/matthew-brett/numpy-wheels-o4pre/build/1.0.3
* https://ci.appveyor.com/project/matthew-brett/scipy-wheels/build/1.0.22

I'm trying builds for v0.3.0 for manylinux, macOS now:

https://travis-ci.org/MacPython/openblas-libs

Broken for manylinux it seems: https://travis-ci.org/MacPython/openblas-libs/jobs/386864579#L541

**Comment by charris** [comment]
Error don't look too bad so far.

**Comment by matthew-brett** [comment]
OpenBLAS issue for compile error : https://github.com/xianyi/OpenBLAS/issues/1586

**Comment by matthew-brett** [comment]
Fix merged for compile error : https://github.com/xianyi/OpenBLAS/pull/1587

Tests for numpy and scipy with OpenBLAS v0.3.0:

* Linux, macOS : https://travis-ci.org/matthew-brett/numpy-wheels/builds/387794420, https://travis-ci.org/matthew-brett/scipy-wheels/builds/387819608 - Scipy failing one test - see https://github.com/scipy/scipy/issues/8326#issuecomment-394415299
* Windows: https://ci.appveyor.com/project/matthew-brett/numpy-wheels-o4pre/build/1.0.7,  https://ci.appveyor.com/project/matthew-brett/scipy-wheels/build/1.0.25

Wheel-building PRs : 

* https://github.com/MacPython/numpy-wheels/pull/28
* https://github.com/MacPython/scipy-wheels/pull/32

**Comment by matthew-brett** [comment]
Oops - segfaults and missing symbols for Scipy / Openblas - see https://mail.python.org/pipermail/scipy-dev/2018-June/022862.html

**Comment by charris** [comment]
@matthew-brett  You have been busy on this, what is the current status?

**Comment by matthew-brett** [comment]
Sorry - comments over on the scipy list, starting with this mail : https://mail.python.org/pipermail/scipy-dev/2018-June/022862.html

Summary - builds with OpenBLAS 0.3.0 are fixed now, for numpy and scipy.

I'm afraid I haven't tested whether it fixes this particular bug, but I would be very surprised if it didn't.

**Comment by charris** [comment]
@matthew-brett  Has `numpy-wheels` been updated to use the new OpenBLAS?

**Comment by matthew-brett** [comment]
Would you mind merging https://github.com/MacPython/numpy-wheels/pull/28? - then it will be updated ...

**Comment by mattip** [comment]
Will this break the bug-fix release process for previous versions? It seems we need to release 1.14.5

**Comment by charris** [comment]
@mattip No, each release has its own branch on `numpy-wheels`.

**Comment by charris** [comment]
@matthew-brett Done, could you make that against master also?

**Comment by charris** [comment]
In fact, there look to be a lot of changes in daily that are not in master.

**Comment by charris** [comment]
We should probably try to set things up so daily can be merged into master on occasion.

**Comment by matthew-brett** [comment]
If you don't mind, I'd like to use Pauli's scheme for scipy, where we use `master` for daily builds (and delete the `daily` branch), and then we branch from `master` to make the release branches.  What do you think?

**Comment by charris** [comment]
That sounds like a great idea. If I diff master and daily, it looks like the only unfortunate difference is that daily still has `windows-wheel-builder`, otherwise a merge would be fine.

**Comment by charris** [comment]
Oh, and @mattip wants to try his hand at the 1.14.5 release, so will need commit privileges.

**Comment by mattip** [comment]
It seems the `clean_code ... ` step is not checking out HEAD on daily builds, [the last run](https://travis-ci.org/MacPython/numpy-wheels/jobs/390900295#L487) seems to have checked out 9e419a035 which is from 8 weeks ago. Is the setup missing a `git pull origin` for numpy somewhere?

**Comment by charris** [comment]
@mattip The last daily build was job 750 , ran with 
```
Processing ./wheelhouse/numpy-1.15.0.dev0+8525d3e-cp27-cp27mu-manylinux1_x86_64.whl
Installing collected packages: numpy
Successfully installed numpy-1.15.0.dev0+8525d3e
NumPy version 1.15.0.dev0+8525d3e
```
So was current. The best way to find these is to look at the history, https://travis-ci.org/MacPython/numpy-wheels/builds .

**Comment by charris** [comment]
Pushing this off to 1.16 just in case it isn't fixed in current OpenBLAS.

**Comment by matthew-brett** [comment]
Yes, I think so.   The current build of numpy dev should have the new OpenBLAS, and does not hang with `python3 -c 'import numpy as np; R = np.array([[1. , 0. , np.nan], [0. , 0 , np.nan], [ 0. , 0, np.nan]]) ;np.linalg.svd(R)'`.  I think the 1.14.x release is using the older numpy-wheels setup, so I guess that will still break, because of the older OpenBLAS.

**Comment by charris** [comment]
I just branched 1.15, so we should have a fixed version out fairly soon.

**Comment by mattip** [comment]
Can confirm this hangs on the 1.14.0 wheel, but passes (raises a `LinAlgError`) with the 1.15.4 wheel. Closing.
